doc/thingpedia-nl-support.md

```diff
@@ -67,10 +154,80 @@ action share(in req status: String #_[prompt="What do you want to post?"])
 ```
 
 By default, the slot filling question will be “What's the value of status?”.
-Now with the prompt provided, Almond will ask “What do you want to post” instead. 
+Now with the prompt provided, Almond will ask “What do you want to post?” instead. 
+
+### Confirmation 
+Before a program is executed, Almond will confirm with the user 
+to make sure it understands the command correctly. 
+The confirmation sentence comes from the `confirmation` annotation of the function.
+Similar to the `utterances` annotation for example commands, 
+use noun phrases for queries and verb phrases in the imperative form for actions. 
+Examples can be found in all the tutorials. 
+
+Note that the confirmation sentence should include all the __required__ input parameter
+with `$param` or `${param}`.
 
 ### Output format
 When a stream or a query is combined with the action `notify`, 
 the returned results will be presented to the user in Almond. 
 By default, Almond will simply list the value of all the output parameters one by one 
-in the order they are declared in the function signature.
\ No newline at end of file
+in the order they are declared in the function signature.
+
+You can customize the output format with the `formatted` annotation.
+The annotation takes a list of Objects. 
+Each Object represents one response and each response will be presented to users in the same order in the list. 
+Four types of responses are supported: `text`, `picture`, `rdl`, and `code`.
+
+#### Text response 
+A text response returns a simple text message.
+It has only one property: `text`. 
+E.g., `{ type="text",text="Works in ${industry}" }` (in [Tutorial 3](/doc/thingpedia-tutorial-linkedin.md)).
+
+####  Picture response: 
+A picture response shows a picture to the user. 
+It has only one property: `url`.
+E.g., `{ type: "picture", url: "${picture_url}" }` (in [Tutorial 3](/doc/thingpedia-tutorial-linkedin.md)).
+
+#### RDL response: 
+An RDL response returns a clickable with a title and a description, suitable for website 
+links and news articles.
+It has three properties: `webCallback` (the link), `displayTitle` (the title), 
+and `displayText` (the description, optional). 
+See [Tutorial 1](/doc/thingpedia-tutorial-nyt.md) for an example of this format type.
+
+#### Customized response
+If you need more control over the output, such as different output based 
+on results, you can choose `code` type and write Javascript code in the `code` property. 
+The code contains a [self-executing anonymous function](https://developer.mozilla.org/en-US/docs/Glossary/IIFE)
+in String format.
+The function will be invoked with three arguments: 
+the result of your function (an object with each parameter as a property), 
+a hint informing you of how the result will be shown to the user, 
+and a `Formatter` object. 
+The function can return a string, a formatted message object 
+(with the same structure as the JSON described here) 
```

Comment:
```suggestion
(with the same structure as the message objects described earlier) 
```

Note that placeholders will **not** be substituted in the value returned by a format function.